core/embed/rust/src/storagedevice/storage_device.rs

```diff
@@ -0,0 +1,690 @@
+const SD_SALT_AUTH_KEY: Field<Vec<u8, SD_SALT_AUTH_KEY_LEN_BYTES>> =
+    Field::public(APP_DEVICE, 0x12).exact();
+const INITIALIZED: Field<bool> = Field::public(APP_DEVICE, 0x13);
+const SAFETY_CHECK_LEVEL: Field<u8> = Field::private(APP_DEVICE, 0x14);
+const EXPERIMENTAL_FEATURES: Field<bool> = Field::private(APP_DEVICE, 0x15);
+
+const SAFETY_CHECK_LEVEL_STRICT: u8 = 0;
+const SAFETY_CHECK_LEVEL_PROMPT: u8 = 1;
+const DEFAULT_SAFETY_CHECK_LEVEL: u8 = SAFETY_CHECK_LEVEL_STRICT;
+const SAFETY_CHECK_LEVELS: [u8; 2] = [SAFETY_CHECK_LEVEL_STRICT, SAFETY_CHECK_LEVEL_PROMPT];
+
+// TODO: somehow determine the DEBUG_MODE value
+const DEBUG_MODE: bool = true;
+const AUTOLOCK_DELAY_DEFAULT: u32 = 10 * 60 * 1000; // 10 minutes
+const AUTOLOCK_DELAY_MINIMUM: u32 = if DEBUG_MODE {
+    10 * 1000 // 10 seconds
+} else {
+    60 * 1000 // 1 minute
+};
+// autolock intervals larger than AUTOLOCK_DELAY_MAXIMUM cause issues in the
+// scheduler
+const AUTOLOCK_DELAY_MAXIMUM: u32 = 0x2000_0000; // ~6 days
+
+const STORAGE_VERSION_CURRENT: u8 = 0x02;
+
+extern "C" fn storagedevice_is_version_stored() -> Obj {
+    let block = || Ok(VERSION.has().into());
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_get_version() -> Obj {
+    let block = || {
+        VERSION
+            .get()
+            .map_or(Ok(Obj::const_none()), |x| (&x as &[u8]).try_into())
+    };
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_set_version(version: Obj) -> Obj {
+    let block = || {
+        VERSION
+            .set(Buffer::try_from(version)?.try_into()?)?
+            .try_into()
+    };
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_is_initialized() -> Obj {
+    let block = || Ok(INITIALIZED.get().unwrap_or(false).into());
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_set_is_initialized(is_initialized: Obj) -> Obj {
+    let block = || INITIALIZED.set(is_initialized.try_into()?)?.try_into();
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_get_rotation() -> Obj {
+    let block = || Ok(ROTATION.get().unwrap_or(0).into());
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_set_rotation(rotation: Obj) -> Obj {
+    let block = || {
+        let rotation = rotation.try_into()?;
+
+        if ![0, 90, 180, 270].contains(&rotation) {
+            Err(Error::ValueError(cstr!("Not valid rotation")))
+        } else {
+            ROTATION.set(rotation)?.try_into()
+        }
+    };
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_get_label() -> Obj {
+    let block = || LABEL.get().map_or(Ok(Obj::const_none()), |x| x.try_into());
+    unsafe { util::try_or_raise(block) }
+}
+
+extern "C" fn storagedevice_set_label(label: Obj) -> Obj {
+    let block = || {
+        let label: StrBuffer = label.try_into()?;
+        LABEL.set(label.try_into()?)?.try_into()
```

Comment:
worth considering something like:

```
// storage_field.rs
pub trait FieldSetSlice<T> {
    fn set_slice(&self, slice: &T);
}

impl FieldSetSlice<str> for Field<String> { ... }
impl FieldSetSlice<[u8]> for Field<Vec<u8>> { ... }
```

then you can use `LABEL.set(label.as_ref())` without the additional `?`. it's a little less try_intos and we avoid the unnecessary `String` or `Vec` middle step

Reply:
```rust
impl FieldGetSet<String> for Field<String> {
    // ...

    fn set(&self, value: String) -> StorageResult {
        self.set_slice(value.as_ref())
    }
}

Reply:
> we avoid the unnecessary String or Vec middle step

Not sure how we could do it, as `String` and `Vec` still need to be in the `set()` signature

> it's a little less try_intos

We would need anyway to copy the `impl<const N: usize> TryFrom<StrBuffer> for String<N>` logic with length validation